userapi/src/main/java/fk/prof/userapi/cache/ZKBasedCacheInfoRegistry.java

```diff
@@ -0,0 +1,264 @@
+        ensureConnected();
+        NodeLoadInfo nodeLoadInfo = readFrom(zkNodesLoadInfoPath, NodeLoadInfo.parser());
+        int profileLoadedCount = nodeLoadInfo.getProfilesLoaded() + 1;
+        CuratorTransactionFinal transaction = curatorClient.inTransaction().setData().forPath(zkNodesLoadInfoPath, buildNodeLoadInfo(profileLoadedCount).toByteArray()).and();
+        if (profileNodeExists) {
+            transaction = transaction.delete().forPath(zkPathForProfile(profileName)).and();
+        }
+        transaction = transaction.create().withMode(CreateMode.EPHEMERAL).forPath(zkPathForProfile(profileName), myResidencyInfoInBytes).and();
+        transaction.commit();
+    }
+
+    /**
+     * Deletes profile residency node for the provided profile name, along with decrementing the owner node's load
+     * Note: This method deletes the node only if it owns it
+     * @param profileName name of the profile
+     * @throws Exception zookeeper exception
+     */
+    private void removeProfileResidencyInfo(AggregatedProfileNamingStrategy profileName) throws Exception {
+        ensureConnected();
+        NodeLoadInfo nodeLoadInfo = readFrom(zkNodesLoadInfoPath, NodeLoadInfo.parser());
+        byte[] newData = buildNodeLoadInfo(nodeLoadInfo.getProfilesLoaded() - 1).toByteArray();
+        curatorClient.inTransaction().setData().forPath(zkNodesLoadInfoPath, newData)
+            .and().delete().forPath(zkPathForProfile(profileName))
+            .and().commit();
+    }
+
+    @VisibleForTesting
+    AutoCloseable getLock() throws Exception {
+        ensureConnected();
+        return new CloseableSharedLock();
+    }
+
+    ConnectionState getState() {
+        return connectionState.get();
+    }
+
+    private class CloseableSharedLock implements AutoCloseable {
+        CloseableSharedLock() throws Exception {
+            if (!interProcessLock.acquire(5, TimeUnit.SECONDS)) {
+                lockAcquireTimeoutCounter.inc();
+                throw new ZkStoreUnavailableException("Could not acquire lock in the given time.");
+            }
+        }
+
+        @Override
+        public void close() throws Exception {
+            interProcessLock.release();
+        }
+    }
+
+    public enum ConnectionState {
+        Disconnected,
+        Connected
+    }
+
+    private void ensureConnected() throws ZkStoreUnavailableException {
+        if(!ConnectionState.Connected.equals(connectionState.get())) {
+            throw new ZkStoreUnavailableException("connection lost recently: " + recentlyZkConnectionLost.get() + ", lastTimeOfLostConnection: " + lastZkLostTime.get().toString());
+        }
+    }
+
+    private NodeLoadInfo buildNodeLoadInfo(int profileCount) {
+        return NodeLoadInfo.newBuilder().setProfilesLoaded(profileCount).build();
+    }
+
+    private boolean sessionIdMatches(AggregatedProfileNamingStrategy profileName) throws Exception {
```

Comment:
i think the name is apt, since its private and its job is literally to check for stale ephemeral nodes. after few change its usage makes it even more clear.